store/json.go

```diff
@@ -109,20 +112,42 @@ func (kvs *jsonFileStore) Flush() error {
 
 // Lock-free flush for internal callers.
 func (kvs *jsonFileStore) flush() error {
-	file, err := os.Create(kvs.fileName)
+	buf, err := json.MarshalIndent(&kvs.data, "", "\t")
 	if err != nil {
 		return err
 	}
-	defer file.Close()
 
-	buf, err := json.MarshalIndent(&kvs.data, "", "\t")
+	dir, file := filepath.Split(kvs.fileName)
+	if dir == "" {
+		dir = "."
+	}
+
+	f, err := ioutil.TempFile(dir, file)
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot create temp file: %v", err)
 	}
 
-	if _, err := file.Write(buf); err != nil {
-		return err
+	tmpFileName := f.Name()
+
+	defer func(){
+		if err != nil {
+			// remove temp file after job is done
+			_ = os.Remove(tmpFileName)
+			f.Close()
+		}
+	}()
+
+	if err = ioutil.WriteFile(tmpFileName, buf, 0); err != nil {
+		return fmt.Errorf("ioutil.WriteFile failed with: %v", err)
 	}
+
+	f.Close()
+
+	log.Printf("renaming tmp file %v to state file", tmpFileName)
+	if err = os.Rename(tmpFileName, kvs.fileName); err != nil {
```

Comment:
The current stdlib implementation seems to have improved since the time the `atomic` package was created, but I took a closer look and one difference is the [usage of the MOVEFILE_WRITE_THROUGH flag on the syscall](https://github.com/natefinch/atomic/blob/master/file_windows.go#L29). It's [documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-movefileexw#parameters) states:


MOVEFILE_WRITE_THROUGH8 (0x8) | The function does not return until the file is actually moved on the disk.Setting this value guarantees that a move performed as a copy and delete operation is flushed to disk before the function returns. The flush occurs at the end of the copy operation.
-- | --

So I'm wondering if this is really atomic. Thoughts?

Reply:
let me check what os.rename implementation is

Reply:
looks like os.replace uses MOVEFILE_REPLACE_EXISTING